File: Algorithms-and-data-structures-master/0.4 Haffman/main.py
from collections import deque

with open("huffman.in", "r") as f:
    n = int(f.readline().strip())

    freq_deq = deque((map(int, f.readline().split())))



second_deq = deque()
result = 0

while len(freq_deq) + len(second_deq) > 1:
    if not second_deq or (freq_deq and freq_deq[0] <= second_deq[0]):
        min1 = freq_deq.popleft()
    else:
        min1 = second_deq.popleft()

    if not second_deq or (freq_deq and freq_deq[0] <= second_deq[0]):
        min2 = freq_deq.popleft()
    else:
        min2 = second_deq.popleft()

    sum_of_mins = min1 + min2
    result += sum_of_mins
    second_deq.append(sum_of_mins)

with open("huffman.out", "w") as f:
    f.write(f"{result}\n")

File: Algorithms-and-data-structures-master/0.4 Haffman/main.cpp
#include <fstream>
#include <deque>
#include <vector>


int main() {
    std::ifstream fin("huffman.in");

    int n;
    fin >> n;

    std::vector<int> frequences(n);

    for (int i = 0; i < n; i++) {
        fin >> frequences[i];
    }
    fin.close();

    std::vector<int> sec_vect;


    long long result = 0;
    int sum_of_mins;
    int min1, min2;

    int index = 0;
    int temp_size;
    while (n + sec_vect.size() > 1) {
        if (sec_vect.empty() || (!sec_vect.empty() and sec_vect[sec_vect.size() - 1] > frequences[index])) {
            min1=frequences[index];
            index++;
            n--;
        }
        else {
            min1 = sec_vect.back();
            sec_vect.pop_back();
        }

        if (sec_vect.empty() || (!sec_vect.empty() and sec_vect[sec_vect.size() - 1] > frequences[index])) {
            min2 = frequences[index];
            index++;
            n--;
        }
        else {
            min2 = sec_vect.back();
            sec_vect.pop_back();
        }

        sum_of_mins = min1 + min2;

        result += sum_of_mins;
        sec_vect.push_back(sum_of_mins);
        temp_size = sec_vect.size();
        if (temp_size>sec_vect[temp_size - 1] > sec_vect[temp_size - 2]) {

        }
    }

    std::ofstream fout("huffman.out");

    fout << result;

    fout.close();
    return 0;
}